src/cata_ui.cpp

```diff
@@ -0,0 +1,1038 @@
+                       max_health( size_x * points_per_char ), current_health( max_health )
+{
+    for( unsigned int i = 0; i < get_rect().size_x; i++ ) {
+        bar_str += "|";
+    }
+}
+
+void health_bar::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    mvwprintz( win, get_ay(), get_ax(), bar_color, bar_str.c_str() );
+}
+
+void health_bar::refresh_bar( bool overloaded, float percentage )
+{
+    bar_str = "";
+    if( overloaded ) {
+        for( unsigned int i = 0; i < get_rect().size_x; i++ ) {
+            bar_str += "*";
+        }
+    } else {
+        for( unsigned int i = 0; i < get_rect().size_x; i++ ) {
+            int char_health = current_health - ( i * points_per_char );
+            if( char_health <= 0 ) {
+                bar_str += ".";
+            } else if( char_health == 1 ) {
+                bar_str += "\\";
+            } else {
+                bar_str += "|";
+            }
+        }
+    }
+
+    static const std::array<nc_color, 7> colors {{
+        c_green,
+        c_ltgreen,
+        c_yellow,
+        c_magenta,
+        c_ltred,
+        c_red,
+        c_ltgray
+    }};
+
+    bar_color = colors[(int) roundf( colors.size() * percentage )];
+}
+
+void health_bar::set_health_percentage( float percentage )
+{
+    bool overloaded = false;
+    if( percentage > 1 ) {
+        current_health = max_health;
+        overloaded = true;
+    } else if( percentage < 0 ) {
+        current_health = 0;
+    } else {
+        current_health = percentage * max_health;
+    }
+
+    refresh_bar( overloaded, percentage );
+}
+
+smiley_indicator::smiley_indicator( int x, int y, ui_anchor anchor) : ui_element(2, 1, x, y, anchor )
+{
+    set_state( neutral );
+}
+
+void smiley_indicator::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    mvwprintz( win, get_ay(), get_ax(), smiley_color, smiley_str.c_str() );
+}
+
+void smiley_indicator::set_state( smiley_state new_state )
+{
+    state = new_state;
+    switch( new_state ) {
+        case very_unhappy:
+            smiley_color = c_red;
+            smiley_str = "D:";
+            break;
+        case unhappy:
+            smiley_color = c_ltred;
+            smiley_str = ":(";
+            break;
+        case neutral:
+            smiley_color = c_white;
+            smiley_str = ":|";
+            break;
+        case happy:
+            smiley_color = c_ltgreen;
+            smiley_str = ":)";
+            break;
+        case very_happy:
+            smiley_color = c_green;
+            smiley_str = ":D";
+            break;
+    }
+}
+
+template<class T>
+tile_panel<T>::tile_panel( size_t size_x, size_t size_y, int x, int y, ui_anchor anchor )
+                       : ui_element( size_x, size_y, x, y, anchor )
+{
+    num_tiles = size_x * size_y;
+    tiles = new T[num_tiles];
```

Comment:
I had some concerns about how vector did it's memory management. (I kind of left it in like this to see what you would say :P).

I was unsure about how ```std::vector```handles polymorphic types. If I have an ```std::vector<T>``` to which I add an object of a T derived type, does it 'slice' the object to only add the ```T``` part? Or does it allocate a new spot in memory and save a pointer to the derived object? 

In java/C# you can do this with the vector equivalents and it does not slice the object. So I'm not sure what C++ does.

But allocating new heap memory takes time, and is not intended (we want a tile panel to hold only 1 type of tile). so I wanted to make sure the object was sliced if it was of a derived type. Previously I used placement new and explicitly called T's copy constructor, but I changed it to array because I figured it probably did the same.